### ISSUE TITLE
Details on training with custom dataset

### ISSUE THREAD

**Issue by shekarneo** [meaten/FlowChain-ICCV2023#7]
Hi,

Thanks for this amazing work, i am trying to train this using custom dataset,
it will be helpful if you provide the steps to prepare the dataset and training information.

**Comment by meaten** [comment]
Hi,

You need raw X-Y positions with frame and tracking IDs for your custom dataset.
We follow the data preprocessing of Trajectron++ as described in `python src/data/TP/process_data.py`.
Feel free to ask me further questions.

**Comment by shekarneo** [comment]
Hi, thanks for the response i have followed and prepared the dataset with frame_num, track_id, x,y and trained with default parameters, i got this results **{'score': 0.06519699238928783, 'ade': 0.06519699238928783, 'fde': 0.06811020451167038, 'nsample': 1.0}** here the score abd ade both are coming same and is there  any data pre-processing step i have to take my data is normalized to be between 0 and 1, when i run test code even i have more data in test set it is always taking one particular file and giving me the results

**Comment by meaten** [comment]
I'm glad to hear that you successfully preprocessed the custom dataset.
The 'score' is currently equal to the 'ade' score. This 'score' is used for hyperparameter tuning using Optuna.
You can change it freely as the mean of 'ade' and 'fde', or your custom metrics.

I couldn't understand the question on data preprocessing.
Could you elaborate more on this?

**Comment by shekarneo** [comment]
Hi

**I couldn't understand the question on data preprocessing.
Could you elaborate more on this?**

i can see the process data code for ETH data set you are multiplying with 0.6 value for test set, is there any specific reason to do that.

and for testing i have more than 30 files of data in test set, but when i run test with visualize it is always takes only one file saving the results.

below is the screen shot

![Image](https://github.com/user-attachments/assets/5825125a-6e24-407c-bbb0-48a8bc564fa1)

**Comment by meaten** [comment]
For multiplying with a 0.6 value, I followed the preprocessing by previous methods on ETH/UCY.

The test code will load the test dataset stored in `{root}/src/data/TP/processed_data/{dataset_name}_test.pkl`.
You should concatenate these 30 data files into one file.

**Comment by shekarneo** [comment]
**The test code will load the test dataset stored in {root}/src/data/TP/processed_data/{dataset_name}_test.pkl.
You should concatenate these 30 data files into one file.**

yes i have done the same even while reading the files while processing i can see the different name as well as i can able to print the different file names in evaluate code, but the output images it is only saving with only one file name.

UPDATE:
using rand=True in main.py i am able to test with different file data
```
data_loader_one_each = unified_loader(
                cfg, rand=True, split="test", batch_size=1)
```

below is the code to read the custom data

```
print("RUNNING for custom data")
for desired_source in ['custom']:
    for data_class in ['train', 'val', 'test']:
        env = Environment(node_type_list=['PEDESTRIAN'], standardization=standardization)
        attention_radius = dict()
        attention_radius[(env.NodeType.PEDESTRIAN, env.NodeType.PEDESTRIAN)] = 3.0
        env.attention_radius = attention_radius

        scenes = []
        data_dict_path = os.path.join(data_folder_name, '_'.join([desired_source, data_class]) + '.pkl')

        for subdir, dirs, files in os.walk(os.path.join(raw_path, desired_source, data_class)):
            for file in files:
                if file.endswith('.json'):
                    input_data_dict = dict()
                    full_data_path = os.path.join(subdir, file)
                    print('Processing custom dataset at', full_data_path)

                    with open(full_data_path, 'r') as f:
                        data = json.load(f)

                    # Prepare data in a similar format as ETH/Stanford
                    data_dict = []
                    for track in data['tracks']:
                        track_id = track
                        for i, frame in enumerate(data['tracks'][track]):
                            frame_id = i
                            if not frame or i % 10 != 0:
                                continue
                            x = frame[0]
                            y = frame[1]
                            data_dict.append([frame_id // 10, track_id, x, y])

                    df = pd.DataFrame(data_dict, columns=['frame_id', 'track_id', 'pos_x', 'pos_y'])

                    df['frame_id'] = pd.to_numeric(df['frame_id'], downcast='integer')
                    df['track_id'] = pd.to_numeric(df['track_id'], downcast='integer')

                    df.sort_values('frame_id', inplace=True)
                    df.to_csv("custom_data.csv", index=False)

                    max_timesteps = df['frame_id'].max()
                    scene = Scene(timesteps=max_timesteps+1, dt=dt, name=file.rstrip('.json'), aug_func=augment if data_class == 'train' else None)

                    for track_id in pd.unique(df['track_id']):
                        node_df = df[df['track_id'] == track_id]
                        node_values = node_df[['pos_x', 'pos_y']].values

                        if node_values.shape[0] < 2:
                            continue

                        new_first_idx = node_df['frame_id'].iloc[0]

                        x = node_values[:, 0]
                        y = node_values[:, 1]
                        vx = derivative_of(x, scene.dt)
                        vy = derivative_of(y, scene.dt)
                        ax = derivative_of(vx, scene.dt)
                        ay = derivative_of(vy, scene.dt)
                        
                        data_dict = {('position', 'x'): x,
                                     ('position', 'y'): y,
                                     ('velocity', 'x'): vx,
                                     ('velocity', 'y'): vy,
                                     ('acceleration', 'x'): ax,
                                     ('acceleration', 'y'): ay}

                        node_data = pd.DataFrame(data_dict, columns=data_columns)
                        # node_data.to_csv("node_data.csv", index=False)
                        node = Node(node_type=env.NodeType.PEDESTRIAN, node_id=track_id, data=node_data)
                        node.first_timestep = new_first_idx

                        scene.nodes.append(node)

                    scenes.append(scene)

        env.scenes = scenes

        if len(scenes) > 0:
            with open(data_dict_path, 'wb') as f:
                dill.dump(env, f, protocol=dill.HIGHEST_PROTOCOL)
```

**Comment by meaten** [comment]
Hi, 

Honestly, I'm unsure why the test code outputs only one example on your custom dataset.
However, the test code tries to visualize the first 10 samples and stops with `break` in Line 150 of `main.py`
If your test dataset has only one example, `len(data_loader_one_each.dataset)` equals 1.

**Comment by shekarneo** [comment]
Thanks, can you give me more details or model stack and default configs details if possible

**Comment by shekarneo** [comment]
One more thing the code is not able to handle the missing data when i replace them with nan, and set **ACCEPT_NAN: True**

**Comment by shekarneo** [comment]
@meaten  is there any information about processing original UCY data set into raw_data format in this repo.

**Comment by meaten** [comment]
Hi, 
`src/data/TP/process_data.py` preprocess the ETH/UCY dataset.
L107 `for desired_source in ['eth', 'hotel', 'univ', 'zara1', 'zara2']:` is the for-loop on each subset.
While the ETH dataset has 'eth' and 'hotel' subsets, UCY dataset has 'univ', 'zara1', and 'zara2'.

**Comment by shekarneo** [comment]
@meaten  i need the information that how the UCY original data set processed into raw_data format

i have downloaded UCY dataset and i can see the data format in zara01 file is **279.000000 -123.000000 0 87.397438 - (2D point, m_id)**   first 2 columns are real world co ordinates (x,y) and in raw_data i can see the format **0.0	1.0	13.4487205051	3.93788669527** here frame_num, track_id, x, y but how this x,y values derived from real world co ordinates

**Comment by meaten** [comment]
I just followed the preprocessing of the previous method.
I think this Github repo https://github.com/Habiba-Amroune/ETH-UCY-Preprocessing is helpful for raw_data into our format

**Comment by shekarneo** [comment]
Thank you
